File: package/main/src/Simple/Array/quickSortSimple.ts
import type { CompareFunction } from "$/array/compareFunction";
import { compareFunctionDefault } from "@/Array/compareFunctionDefault";
import { quickSort } from "@/Array/quickSort";

/**
 * Quick sort implementation for arrays
 * @template T Type of array elements
 * @param {T[]} array Input array to sort
 * @param {(a: T, b: T) => number} compareFunction Function to determine sort order
 * @param {number} startID Starting index for sorting
 * @param {number} endID Ending index for sorting
 * @returns {T[]} Sorted array
 * @example quickSort([1, 3, 2, 4, 5], (a, b) => a - b); // [1, 2, 3, 4, 5]
 */
export const quickSortSimple = <T>(
  array: T[],
  compareFunction: CompareFunction<T> = compareFunctionDefault<T>,
  startID = 0,
  endID: number = array.length - 1,
): T[] => {
  let localStartID = startID;
  let localEndID = endID;
  if (
    localStartID < 0 ||
    localStartID >= array.length ||
    localStartID > localEndID
  ) {
    localStartID = 0;
  }
  if (localEndID < 0 || localEndID >= array.length) {
    localEndID = array.length - 1;
  }
  return quickSort(array, compareFunction, localStartID, localEndID);
};

File: package/umt_python/src/simple/array/quick_sort_simple.py
from collections.abc import Callable
from typing import TypeVar

from src.array.quick_sort import quick_sort

T = TypeVar("T")
CompareFunction = Callable[[T, T], int | float]


def quick_sort_simple(
    array: list[T],
    compare_function: CompareFunction[T] | None = None,
    start_id: int = 0,
    end_id: int | None = None,
) -> list[T]:
    """
    Quick sort implementation for arrays with simple validation logic.

    Args:
        array: Input array to sort.
        compare_function: Function to determine sort order.
        start_id: Starting index for sorting (default 0).
        end_id: Ending index for sorting (default array.length - 1).

    Returns:
        Sorted array.
    """
    local_start_id = start_id

    local_end_id = len(array) - 1 if end_id is None else end_id

    if (
        local_start_id < 0
        or local_start_id >= len(array)
        or local_start_id > local_end_id
    ):
        local_start_id = 0

    if local_end_id < 0 or local_end_id >= len(array):
        local_end_id = len(array) - 1

    return quick_sort(array, compare_function, local_start_id, local_end_id)
